REP-3222 Upgrade to Scala v2.11

wdschei opened the pull request:
**NOTE:** The old yammer `metrics.gauge` did not care if there was already a metric with the same name, it just passively went on with life. The new `dropwizard`/`codahale`/`nl.grons` version now throws a `RuntimeException`. Currently this is being handled by logging the fact at the `INFO` level. This could be ignored or made less verbose. However, this is the exceptional case and should not usually be hit outside of testing. The other alternative would be to add a deconstructor to each of the pools and have them usregister the gauges that were registered upon helper object instantiation.

dmnjohns commented on core/src/main/scala/com/rackspace/com/papi/components/checker/Validator.scala:
```diff
@@ -25,18 +25,27 @@ import javax.xml.transform.dom._
+
 object Validator {
+  /** The application wide metrics registry. */
+  val metricRegistry = new com.codahale.metrics.MetricRegistry()
```
Would it make sense to break this out into its own object? Some sort of application context.

It seems odd to keep the application-wide metrics registry in the Validator object.

wdschei replied:
The Validator is the application context for API-Checker.

wdschei replied:
At the same time, we could explicitly break it out into its own object.
We just need to move the Instrumented trait to wherever it goes.

dmnjohns replied:
I'm not sure I'd call it the application context. It's the entry-point, sure, but from what I can tell these would be the first universally accessible fields.

Personally, I would put this somewhere else. Of course, you probably know the api-checker codebase better than I do, so I won't complain if you think this is the right place for it to be.

wdschei replied:
What do we want to name the new context?

- ValidatorContext
- CheckerContext
- Other???

dmnjohns replied:
That's a great question... for someone else. I suck at names.

Either of those sound fine.

RackerWilliams replied:
To break what into its own object.  The metricRegistry?  The Validator object is a singleton, and I'm assuming we want the registry to also be a singleton and there's a 1:1 relationship...right?

wdschei replied:
Ok. That does make sense. I'll leave it in the Balidator.

dmnjohns replied:
A 1:1 relationship to what? In Repose, for example, you may have multiple validators, but you'll still only be getting (and probably wanting) one metric registry to handle all api-checker metrics. I would expect the metricRegistry to be in an object, just perhaps not this object.

Putting the metricRegistry in the Validator object certainly works. It just seems to me like the function of the metricRegistry is not strictly tied to a Validator. Let me put it this way: if I wanted to add some metric based on a URLFail, would I want to import the Validator object and class? Importing the Validator object and class would provide a lot of code in scope that isn't needed when all I want to do it record some metric. Granted, in this case you can simply mix-in the Instrumented trait, which is great for handling this type of thing. Still, I would personally break out the metricRegistry for the sake of clarity and code organization, but I understand that it will work as-is. As this is @RackerWilliams brainchild, I defer to him.

dmnjohns commented on core/src/main/scala/com/rackspace/com/papi/components/checker/Validator.scala:
```diff
@@ -98,8 +107,8 @@ class Validator private (private val _name : String, val startStep : Step, val c
-  private val objectName = new ObjectName("\"com.rackspace.com.papi.components.checker\":type=\"Validator\",scope=\""+
-                                 name+"\",name=\"checker\"")
+  private val objectName = new ObjectName("com.rackspace.com.papi.components.checker:type=Validator,scope="+
+                                 name+",name=checker")
```
I assume the " characters are no longer necessary with the updated library, but that the output is the same.

wdschei replied:
They weren't supposed to have the quotes. I'm not sure why he added them in originally.

dmnjohns replied:
Got it. Are quotes valid characters, though? Or, I guses I should ask, what are these strings used for? If it's just for internal stuff, whatever. If it's user-facing, then a backwards incompatible change makes me a little nervous.

wdschei replied:
These are just the key names to look up in the JMX Registry. So yes, quotes are valid characters and they were showing up in the actual names. From what I can tell, the intent was to identify make them Strings. However they are all Strings and the quotes were included in the String.

I updated it for consistency with all other registered JMX items which did not have them.

wdschei replied:
They are public facing, but only used in the lookup of the metrics in JMX. That said, there may be someone that is using them. At the same time this is a major version with other breaking changes too.

dmnjohns replied:
Fair enough. Major version shifts for the win.

RackerWilliams replied:
The quotes were there because that's what the metrics library did with the rest of its object...otherwise you won't find all the metrics in the same place ...I'll check to see if it's still doing that.  If it is, you probably want to add them back.

wdschei replied:
The quotes are not in the new library.

dmnjohns commented on core/src/main/scala/com/rackspace/com/papi/components/checker/Validator.scala:
```diff
@@ -154,16 +163,20 @@ class Validator private (private val _name : String, val startStep : Step, val c
-  private val failMeter = metrics.meter(FAIL_METER_NAME, FAIL_METER_EVENT, name)
-  private val failGauge =  metricsRegistry.newGauge(getClass, FAIL_RATE_NAME, name,
-                                                   (new ValidatorFailGauge(timer, failMeter)))
+  private val timer = metrics.timer(name, TIMER_NAME)
```
Did the parameter order get switched with the version update? I assume they did so for the sake of consistency, but how annoying!

wdschei replied:
The parameters are a varagrs list of `Strings`. I changed the order for consistency with the other metrics that now use the `MetricRegistry.name()` to construct the key.
I'm not sure why the `timer` method doesn't use it too.

dmnjohns replied:
Ah, I see. So just change the timer method to use it, for consistency?

wdschei replied:
The values are passed in and then become a dot delimited String that is used as the key name for the metric in the registry. With the new library, some of the items utilize the new `MetricRegistry.name()` for the name and some still use the old varargs way. This change was for consistency in the names actually being registered.

dmnjohns replied:
The change makes sense, and I appreciate the consistency.

Here's where I'm getting lost: I pulled this from the nl-grons metrics-scala 3.5.2 jar.

```
package nl.grons.metrics.scala
class MetricBuilder(val baseName : nl.grons.metrics.scala.MetricName, val registry : com.codahale.metrics.MetricRegistry) extends scala.AnyRef {
  def gauge[A](name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ })(f : => A) : nl.grons.metrics.scala.Gauge[A] = { /* compiled code */ }
  def cachedGauge[A](name : scala.Predef.String, timeout : scala.concurrent.duration.FiniteDuration, scope : scala.Predef.String = { /* compiled code */ })(f : => A) : nl.grons.metrics.scala.Gauge[A] = { /* compiled code */ }
  def counter(name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ }) : nl.grons.metrics.scala.Counter = { /* compiled code */ }
  def histogram(name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ }) : nl.grons.metrics.scala.Histogram = { /* compiled code */ }
  def meter(name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ }) : nl.grons.metrics.scala.Meter = { /* compiled code */ }
  def timer(name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ }) : nl.grons.metrics.scala.Timer = { /* compiled code */ }
  protected def metricNameFor(name : scala.Predef.String, scope : scala.Predef.String = { /* compiled code */ }) : scala.Predef.String = { /* compiled code */ }
}
```

It looks to me like the parameters are not varargs anymore, and that MetricRegistry.name(...) would work as the parameter to the timer method. Am I looking at the wrong thing?

dmnjohns replied:
Apparently I was mistaken on this count. @wdschei pointed out that MetricRegistry.name(...) is called by the multi-parameter versions of the MetricBuilder functions. So the best thing to do may be to use those functions without using the MetricRegistry.name(...) function.

wdschei replied:
After looking at the actual code that was underlying these, I just dumped it all together in favor of using the pattern:

- `MetricRegistry.name(getRegistryClassName(getClass), SCOPE, NAME)`

dmnjohns commented on core/src/main/scala/com/rackspace/com/papi/components/checker/Validator.scala:
```diff
@@ -154,16 +163,20 @@ class Validator private (private val _name : String, val startStep : Step, val c
+      new ValidatorFailGauge(timer, failMeter)
+    )
+  } recover {
+    case e: RuntimeException => logger.info("Problem adding new Fail gauge metric.", e)
```
Odd exception to react to. It's more specific than Exception, but more generic than basically any other exception. Is there a reason to use RuntimeException specifically? Scala treats every exception as an unchecked exception, after all.

If you just want a catch-all (i.e., if registering this metric is not vital to the application so you just want to give it a try) perhaps something like the following is clearer?
```
case e: _ => logger.info("Problem adding new Fail gauge metric.", e)
```

dmnjohns replied:
It may be appropriate to log at the WARN level. I'm not sure how api-checker treats metrics, but this does seem like an exceptional case.

wdschei replied:
That works. `RuntimeException` was just the exception that was being thrown when the metric already existed.

wdschei replied:
This was only happening during testing because the new registry would sometimes hang around in between tests.

wdschei replied:
In the old one, it would just passively use the already existing metric if it was there. So now we are at least logging that it happened.

wdschei replied:
Should I log the `e.getmessage` at info and then the actual exception trace at debug?

dmnjohns replied:
Actually, for the exact reason you mentioned, I'd just log everything at warn. It would be good to know if we expected to create a new metric, but ended up colliding with an existing metric. I wouldn't think that to be expected behavior, and a WARN might get someone to actually take a look at it. The naughty case won't prevent the program from doing its thing, but it might give you unexpected metrics.

wdschei replied:
Cool. I'll bump up the level.

RackerWilliams replied:
I would also bump this up to a WARN.  Are we sure that things are still passively being reused??  How did you confirm that?

wdschei replied:
If they weren't being used, then the tests would fail.

wdschei replied:
This was cleaned up by getting rid of the `nl.grons` library and just using the CodaHale/DropWizard library directly.

dmnjohns commented on core/src/main/scala/com/rackspace/com/papi/components/checker/util/XPathExpressionPool.scala:
```diff
@@ -38,8 +40,16 @@ object XPathExpressionPool extends Instrumented {
+      idleGauges :+ metrics.gauge("Idle", expression+" ("+version+")")(pool.getNumIdle)
+    } recover {
+      case e: RuntimeException => logger.info("Problem adding new Idle gauge metric.", e)
+    }
```
I don't love these Try blocks everywhere. I get why we're using them, and it's better than not using them, but man does it make things ugly and repetitive.

Perhaps you can write a function to handle this logic somewhere.

RackerWilliams replied:
...or better yet, this is begging for a macro :-)

dmnjohns replied:
Oh, macros! I still haven't used them, so they didn't come to mind, but that seems like a great idea!

dmnjohns commented:
Despite the novel written in comments, this is good stuff. Even now, there's nothing wrong that would prevent me from merging this.

Let me know once you're done making any changes you feel are appropriate. I'll look over anything you add and then merge this.

wdschei commented:
Cool. I'll start hitting those now.

RackerWilliams commented on core/src/main/scala/com/rackspace/com/papi/components/checker/util/IdentityTransformPool.scala:
```diff
@@ -29,8 +32,16 @@ object IdentityTransformPool extends Instrumented  {
+  Try {
+    metrics.gauge("Active")(numActive)
+  } recover {
+    case e: RuntimeException => logger.info("Problem adding new Active gauge metric.", e)
```
Same here WARN...and confirm that the gauge is still functional afterwards.  Because I'm afraid this will happen not just during testing.  Is it possible to happen with multiple validators??

wdschei replied:
This was extracted out to the trait.

RackerWilliams commented on core/src/test/scala/com/rackspace/com/papi/components/checker/handler/InstrumentedHandlerSuite.scala:
```diff
@@ -86,7 +86,7 @@ class InstrumentedHandlerSuite extends BaseValidatorSuite {
 
   def getStepObjectName (stepId : String) : ObjectName = {
-    new ObjectName("\"com.rackspace.com.papi.components.checker.handler\":type=\"InstrumentedHandler\",scope=\"MyInstTestValidator\",name=\""+stepId+"\"")
+    new ObjectName("com.rackspace.com.papi.components.checker:name=com.rackspace.com.papi.components.checker.handler.InstrumentedHandler.MyInstTestValidator."+stepId)
```
Again, if the rest of the objects are using quotes, please add them back in.

wdschei replied:
The new metrics library does not operate the same as the old. The quotes do not exist in the new one.

RackerWilliams commented:
I'm also seeing a lot of deprecation warnings.  You will address these yes?

````
[WARNING] /Users/jorgew/projects/api-checker/core/src/main/scala/com/rackspace/com/papi/components/checker/Config.scala:35: method tpe in trait AnnotationApi is deprecated: Use `tree.tpe` instead
[WARNING]     universe.typeOf[Config].members.filter(i => i.annotations.map(a => a.tpe).contains(affectsCheckerType))
[WARNING]                                                                          ^
[WARNING] /Users/jorgew/projects/api-checker/core/src/main/scala/com/rackspace/com/papi/components/checker/util/TransformPool.scala:33: class LinkedList in package mutable is deprecated: Low-level linked lists are deprecated due to idiosyncrasies in interface and incomplete features.
[WARNING]   private val activeGauges = new LinkedList[Gauge[Int]]
[WARNING]                                  ^
[WARNING] /Users/jorgew/projects/api-checker/core/src/main/scala/com/rackspace/com/papi/components/checker/util/TransformPool.scala:34: class LinkedList in package mutable is deprecated: Low-level linked lists are deprecated due to idiosyncrasies in interface and incomplete features.
[WARNING]   private val idleGauges = new LinkedList[Gauge[Int]]
[WARNING]                                ^
[WARNING] /Users/jorgew/projects/api-checker/core/src/main/scala/com/rackspace/com/papi/components/checker/util/TransformPool.scala:40: object LinkedList in package mutable is deprecated: Low-level linked lists are deprecated.
[WARNING]       activeGauges :+ metrics.gauge("Active", Integer.toHexString(templates.hashCode()))(pool.getNumActive)
````

wdschei commented:
To fix the deprecation warnings is going to be fairly invasive. I wanted to make that a different PR.

wdschei commented:
I dug deep into the code of the replacement libraries for the old `com.yammer.metrics`. The new `com.codahale.metrics` is almost fully featured but is missing a few things that were deemed unnecessary by Coda, so many are using the `nl.grons.metrics` to fill in the gaps. That is why I also originally used the latter as it had a few things that looked useful and it also had pretty good instructions. As I combed though the code, the few pieces we were actually using were minor. I collapsed and extracted out what we needed. Specifically the new `Instrumented` trait has been fully fleshed based on the instructions for using `nl.grons.metrics` and the collapsing of several of its parent traits.

I think I also might have figured out where the `"` were coming from in the old library. There may have been something being interpreted as a JMX Registry Name pattern, so it would quote it to escape the pattern. If this was the case, then the new libraries were updated to not escape our names. The new library also did not publish to JMX in the same way and was nowhere near as clean to view as the old. I also dug into this and ended up creating our own JMX Object Name Factory. This is used by the new libraries `JmxReporter` to create a similar experience to the old output. It is not exactly the same, but it is close.

By leaving the `nl.grons.metrics` library completely, we got back to the old add if new or get if existing behavior of the old libraries. This allowed me to get rid of all the `Try/recover` repetitiveness. I did need to extract a little bit of the new library to get this fully featured for `Gauge`'s. As part of the extraction, I moved the logged message out to the `trait` as well.

the new `Instrumented` `trait` also provides a clean way to get the properly truncated class name for the `MetricRegistry.name()` call provided by the new libraries.

RackerWilliams commented on core/src/main/scala/com/rackspace/com/papi/components/checker/util/Instrumented.scala:
```diff
@@ -0,0 +1,62 @@
+      Some(buildMetric())
+    } catch {
+      case e: IllegalArgumentException =>
+        logger.warn("Failed to add new metric. Reason: {}", e.getMessage)
```
Were you able to confirm that the metric is used properly when you hit this exception?  Or do you not hit this anymore, given the changes you've made?

wdschei replied:
The latest update makes it so `guard()` can't be called outside of `gaugeOrAdd()`. The latter ensures that either a new `Gauge` was added or the existing one is returned. This is the behavior of the old libraries. So yes we still hit the exception, and yes we do return the already registered `Gauge`.

RackerWilliams commented:
Okay to move deprecation warnings to another PR.  This is looking good.  Thanks for the attention to detail.

RackerWilliams commented:
:+1:  I'm okay with this going through.  @dmnjohns if you're also cool with it, please merge.

dmnjohns commented:
I took another look, and it mostly seems sensible if a bit complex.

If @RackerWilliams is happy, I'm happy. Merge incoming.
